snapshots/erofs/erofs.go

```diff
@@ -0,0 +1,622 @@
+			if os.IsNotExist(err) {
+				log.G(ctx).Infof("loop device for snapshot %s not found, preparing", s.ParentIDs[i])
+				loopDevice, err = o.loopDriver.Prepare(context.Background(), filepath.Join(o.plainPath(s.ParentIDs[i]), "layer.erofs"))
+			} else if err != nil {
+				log.G(ctx).Infof("loop device get failed: %v", err)
+			}
+			options = append(options, fmt.Sprintf("device=%s", loopDevice))
+		}
+
+		lowerMount := mount.Mount{
+			Type:    "erofs",
+			Source:  mergedLoop,
+			Options: options,
+		}
+		if err := os.MkdirAll(lowerDir, 0777); err != nil {
+			log.G(ctx).Errorf("failed to prepare merged dir %v for upper: %v", lowerDir, err)
+			return []mount.Mount{
+				{
+					Source: o.plainPath(s.ID),
+					Type:   "bind",
+					Options: append(options,
+						"rbind",
+						"rw",
+					),
+				},
+			}
+		}
+		if err := lowerMount.Mount(lowerDir); err != nil {
+			log.G(ctx).Errorf("failed to mount merged dir %v for upper: %v", lowerDir, err)
+			return []mount.Mount{
+				{
+					Source: o.plainPath(s.ID),
+					Type:   "bind",
+					Options: append(options,
+						"rbind",
+						"rw",
+					),
+				},
+			}
+		}
+	} else if err != nil {
+		log.G(ctx).Errorf("failed to stat merged dir for upper: %v", err)
+		return []mount.Mount{
+			{
+				Source: o.plainPath(s.ID),
+				Type:   "bind",
+				Options: append(options,
+					"rbind",
+					"rw",
+				),
+			},
+		}
+	}
+
+	options = append(options,
+		fmt.Sprintf("workdir=%s", o.workPath(s.ID)),
+		fmt.Sprintf("upperdir=%s", o.upperPath(s.ID)),
+	)
+	options = append(options, fmt.Sprintf("lowerdir=%s", lowerDir))
+	return []mount.Mount{
+		{
+			Type:    "overlay",
+			Source:  "overlay",
+			Options: options,
+		},
+	}
+}
+
+func (o *snapshotter) mergeDir(ctx context.Context, s storage.Snapshot, path string) error {
+	mergeTarget := filepath.Join(path, "merged.erofs")
+	args := []string{
+		mergeTarget,
+	}
+	for i := range s.ParentIDs {
+		args = append(args, filepath.Join(o.plainPath(s.ParentIDs[i]), "layer.erofs"))
+	}
+	log.G(ctx).Infof("merging %v layers with command: %v", mergeTarget, args)
+	output, err := exec.Command("mkfs.erofs", args...).CombinedOutput()
+	if err != nil {
+		return errors.Wrapf(err, "failed to merge erofs:\n%s", string(output))
+	}
+	return nil
+}
+
+func (o *snapshotter) umount(ctx context.Context, target string) error {
+	args := []string{
+		"-T",
+		target,
+		"-P",
+		"-o",
+		"SOURCE",
+	}
+
+	output, err := exec.Command("findmnt", args...).CombinedOutput()
+	if err != nil {
+		log.G(ctx).Warnf("umount: failed to findmnt target %s: %v", target, err)
+	}
+	if err := mount.Unmount(target, 0); err != nil {
+		return err
+	}
+	loops := strings.Split(string(output), "\"")
+	if len(loops) < 2 {
+		log.G(ctx).Infof("umount: findmnt result is: %v", output)
+		return nil
+	}
+	loop := loops[1]
+	log.G(ctx).Debugf("umount: trying to ensure %s is cleaned on %s", loop, target)
+	for i := 0; i < 30; i++ {
```

Comment:
Why 30?

Reply:
this is an empirical number. In production we encounterd scenario where some pending write in buffer which needs to be synced before umounting the sparsefile, so we poll loop device backend_file in sparsefile umount to ensure device is released and atomic.